src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -206,44 +206,111 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
+                    // it and let the controller decide what to do.  (For example, it will
+                    // pick the one that has the best MRU index).
+                    bestItems.Add(pair);
+
+                    if (highestMatchPriorityInBest < pair.item.Rules.MatchPriority)
                     {
-                        // This item is strictly better than the best items we've found so far.
-                        bestItems.Clear();
-                        bestItems.Add(pair);
+                        highestMatchPriorityInBest = pair.item.Rules.MatchPriority;
+                        // No need to remove items with lower MatchPriority from similarItemsWithHigerMatchPriority list,
+                        // we will only add ones with higher value at the end.
                     }
-                    else if (comparison == 0)
+                }
+                else if (comparison < 0)
+                {
+                    // This item is strictly better than the best items we've found so far.
+                    // However, if it's only better in terms of case-sensitivity, we'd like 
+                    // to save the prior best items and consider their MatchPriority later.
+                    similarItemsWithHigherMatchPriority.Clear();
+
+                    if (filterTextContainsNoUpperLetters &&
+                        onlyDifferInCaseSensitivity &&
+                        highestMatchPriorityInBest > pair.item.Rules.MatchPriority) // don't add if this item has higher MatchPriority than all prior best items
                     {
-                        // This item is as good as the items we've been collecting.  We'll return 
-                        // it and let the controller decide what to do.  (For example, it will
-                        // pick the one that has the best MRU index).
-                        bestItems.Add(pair);
+                        similarItemsWithHigherMatchPriority.AddRange(bestItems);
                     }
+
+                    bestItems.Clear();
+                    bestItems.Add(pair);
+                    highestMatchPriorityInBest = pair.item.Rules.MatchPriority;
+                }
+                else
+                {
                     // otherwise, this item is strictly worse than the ones we've been collecting.
-                    // We can just ignore it.
+                    // However, if it's only worse in terms of case-sensitivity, we'd like 
+                    // to save it and consider its MatchPriority later.
+                    if (filterTextContainsNoUpperLetters &&
+                        onlyDifferInCaseSensitivity &&
+                        pair.item.Rules.MatchPriority > highestMatchPriorityInBest)  // don't add if this item doesn't have higher MatchPriority
+                    {
+                        similarItemsWithHigherMatchPriority.Add(pair);
+                    }
+                }
```

Comment:
i think this logic can be cleaned up by first continuing at the end of the `comparison == 0` branch.  then, in the part that follows, just do the `if (filterText ... && .. && ...)` check once, and then decide waht to do in the comparison `>` or `<` cases.

Reply:
Not sure what exactly you are suggesting, probably my mind is just stuck with he way I implemented it. Could you please provide a little more details? :)

Reply:
ping~

Reply:
n/m.  i see the if checks are different now :)

Reply:
taking offline.